base/genimport/importer.go

```diff
@@ -269,14 +272,120 @@ func removeAllFilesInDirExcept(o *Output, dir string, except_list []string) {
+		[]byte("module gomacro.imports/" + pkgpath + "\n"),
+	}
+
+	if localpath, ok := local(o, pkgpath); ok {
+		newLines = append(newLines, goModReplacementDirectives(o, localpath)...)
+	}
+
+	newGoModText := bytes.Join(newLines, nil)
+
 	gomod := paths.Subdir(dir, "go.mod")
-	err := ioutil.WriteFile(gomod, []byte("module gomacro.imports/"+pkgpath+"\n"), os.FileMode(0644))
+	err := ioutil.WriteFile(gomod, newGoModText, os.FileMode(0o644))
 	if err != nil {
 		o.Errorf("error writing file %q: %v", gomod, err)
 	}
+
 	return gomod
 }
 
+// local returns a localized path for the package if it is local, as well as a
+// boolean indicating if it was found
+func local(o *Output, pkgpath string) (string, bool) {
+	cwd, err := os.Getwd()
+	if err != nil {
+		o.Errorf("error finding cwd: %v", err)
+		return "", false
+	}
+	// Hacky way of determining if we're running gomacro from a directory that
+	// likely shares a go.mod file with the pkgpath. It assumes the both the
+	// cwd and module location will include the pattern 'owner/project', which
+	// is often true.
+	// TODO: find a better way to determine if it's local
+	modParts := strings.Split(pkgpath, "/")
+	if len(modParts) > 2 &&
+		strings.Contains(strings.ToLower(cwd), strings.ToLower(strings.Join(modParts[1:3], "/"))) {
+		return cwd, true
+	}
+
+	return "", false
+}
+
+// goModReplacementDirectives will create the replacement directives associated
+// with a module that can be found locally, for the purpose of use in a
+// gomacro.imports mod file.
+func goModReplacementDirectives(o *Output, localpath string) [][]byte {
+	goModDir, goModFile, err := findGoMod(localpath)
+	if err != nil {
+		o.Errorf("error finding go.mod: %v", err)
+		return [][]byte{}
+	}
+
+	origGoModText, err := ioutil.ReadFile(goModFile)
+	if err != nil {
+		o.Errorf("error reading go.mod: %v", err)
+		return [][]byte{}
+	}
+	origLines := bytes.SplitAfter(origGoModText, []byte("\n"))
+
+	// Figure out current module name
+	// TODO: Could we use github.com/golang/mod to parse this?
+	var moduleName string
+	for _, line := range origLines {
+		if bytes.HasPrefix(line, []byte("module")) {
+			// TODO: is the module line guaranteed to have this syntax?
+			moduleName = string(bytes.Fields(line)[1])
+			break
+		}
+	}
+	if moduleName == "" {
+		o.Errorf("go.mod had no module line")
+	}
+
+	// Replace the current module with the right directory
+	replacementDirectives := [][]byte{
+		[]byte("replace "+moduleName+" => "+goModDir+"\n"),
+	}
+
+	// Copy over the replaces, since they only apply when in the main module
+	for _, line := range origLines {
+		// TODO: this doesn't handle block syntax e.g. replace ( ... )
+		// TODO: this doesn't work if the replace directives found are themselves localized
+		if bytes.HasPrefix(line, []byte("replace")) {
+			replacementDirectives = append(replacementDirectives, line)
+		}
+	}
+	return replacementDirectives
+}
+
+var noGoModOnPath = errors.New("no go.mod file on path")
+
```

Comment:
As an [alternative to the findGoMod code below](https://github.com/ldez/gomoddirectives/blob/701dfb75f0735a81975ab3cb187e34920f4d7dbf/module.go#L13-L47):
```
type modInfo struct {
	Path      string `json:"Path"`
	Dir       string `json:"Dir"`
	GoMod     string `json:"GoMod"`
	GoVersion string `json:"GoVersion"`
	Main      bool   `json:"Main"`
}

// GetModuleFile gets module file.
func GetModuleFile() (*modfile.File, error) {
	// https://github.com/golang/go/issues/44753#issuecomment-790089020
	cmd := exec.Command("go", "list", "-m", "-json", "-f", "{{.GoMod}}")

	raw, err := cmd.CombinedOutput()
	if err != nil {
		return nil, fmt.Errorf("command go list: %w: %s", err, string(raw))
	}

	var v modInfo
	err = json.Unmarshal(raw, &v)
	if err != nil {
		return nil, fmt.Errorf("unmarshaling error: %w: %s", err, string(raw))
	}

	if v.GoMod == "" {
		return nil, errors.New("working directory is not part of a module")
	}

	raw, err = ioutil.ReadFile(v.GoMod)
	if err != nil {
		return nil, fmt.Errorf("reading go.mod file: %w", err)
	}

	return modfile.Parse("go.mod", raw, nil)
}
```

Reply:
Hm. I'm not sure how I feel about shelling out a command here, especially as we're shelling to go from go... I wonder if I can just leverage the library it's built on though...

Reply:
I don't think it's a big problem to shell out to `go list`, but I think what we have is simple enough that we may as well avoid it.

Reply:
I've looked into it more and this pattern is quite common, so I think I've softened my aversion to shelling out to go. And it does resolve the problem around the naming convention.